### PR TITLE
Provided RW impl for fixed arrays.

### PR DESCRIPTION
**Pull request by Frostie314159**
This PR makes reading and writing possible for fixed length arrays of any type implementing `TryFromCtx` or `TryIntoCtx`.

### CODE REVIEW

**Comment by Frostie314159** [comment]
I finally managed to implement this without the whole merge conflict fiasko and as a little bonus made it work for every type. I'm aware, that this uses unsafe, but I think my intuition, that unsafe is necessarily bad was wrong. If we just use it sparingly and with safety comments we should be good to go.

File: src/ctx.rs
Comment (m4b) [inline src/ctx.rs:791]: this is not actually safe; `*element` line below will call the drop function on `T`, this is a typical pitfall of unsafe rust and using `mem::zeroed`, and hard to catch.

if this were e.g., a String, or some other complex object, it could theoretically segfault during it's drop routine

You have two options:
1. if you want to stick with unsafe, the proper solution is to use `mem::MaybeUninit`, with a write-only initialization of an array being the classic example/usecase.
2. if you don't want to stick with unsafe, i believe you could add a default bound to T, and then do, e.g.: `let mut buf: [T; N] = [T::default(); N]`

I'm ok with MaybeUninit approach, as it is less burdensome on the implementator to require default on their type, however we have to be extra careful during review that it is safe + sound :)

**Review by m4b — CHANGES_REQUESTED** [review]
need to address unsafe use of zeroed as per my comment

File: src/ctx.rs
Comment (Frostie314159) [reply src/ctx.rs:791]: I'm currently working on doing this with `MaybeUninit`, but I don't know how to turn `[MaybeUninit<T>;N]` into `[T;N]` due to the layout guarantees, transmute would be an option.

Comment (m4b) [reply src/ctx.rs:791]: Somewhat related, a fully safe impl that requires a SizeWith bound (e.g., no sizes depending on the ctx) which will rule out dynamically sized things like slices would look something like:
```rust
impl<
        'a,
        Ctx: Copy,
        T: TryFromCtx<'a, Ctx, Error = error::Error> + SizeWith<Ctx>,
        const N: usize,
    > TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let nbytes_required = N * T::size_with(&ctx);
        if nbytes_required > src.len() {
            panic!();
        }
        let buf: [T; N] =
            std::array::from_fn(|i| src.pread_with(i * T::size_with(&ctx), ctx).unwrap());

        Ok((buf, nbytes_required))
    }
}
```

I think the use of unsafe and `MaybeUninit` though is actually justified here, since your impl _is_ the "initializer"; again though we must be very careful during review we don't introduce unsoundness in this part of the library

Comment (Frostie314159) [reply src/ctx.rs:791]: The issue is, that for one we are not handling errors properly, which will be hard, without using try_from_fn which is unstable and on the other hand can't reasonably introduce a Default bound to all types T.

Comment (Frostie314159) [reply src/ctx.rs:791]: I'm currently at an implementation, which looks like this:
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;

        // This is taken from the implementation of `MaybeUninit::uninit_array`, which is currently unstable.
        // SAFETY: An uninitialized `[MaybeUninit<_>; LEN]` is valid.
        let mut buf = unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() };

        for element in buf.iter_mut() {
            *element = MaybeUninit::new(src.gread_with::<T>(&mut offset, ctx)?);
        }

        Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
    }
}
```
The array initialization was taken from `MaybeUninit::uninit_array`, which is unstable. This avoids having to introduce a `Copy` bound on T.

Comment (m4b) [reply src/ctx.rs:791]: This seems like the correct approach, although if we did add a copy bound I don’t think we need any unsafe at all.

One thing that does worry me is that if gread fails in that loop, any objects which got put in the array from earlier in the loop before failure, are never dropped. This is maybe ok but it does worry me a bit. There’s a way to fix this but it’s not great (reiterate back up to the failing index and assume_init it and that handles the drop as far as I know)

Comment (m4b) [reply src/ctx.rs:791]: Ah, i see what you mean now; yes we still need to `MaybeUninit` the array itself, which requires unsafe, if we have `Copy` bound; but we'd still have unsafe; in order to do full unsafe requires `Default`; i think what you've written is ok, though a `Copy` bound would ensure the type is "safer", but not quite enough.

Comment (m4b) [reply src/ctx.rs:791]: Here's another version with only 1 unsafe block, although there's a good chance the array is double-initialized, i.e., first from `from_fn` (likely a memset) then the pread loop, (would need to check assembly):
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        use std::mem::MaybeUninit;

        let mut buf: [MaybeUninit<T>; N] = std::array::from_fn(|_| MaybeUninit::uninit());

        for element in buf.iter_mut() {
            *element = MaybeUninit::new(src.gread_with::<T>(&mut offset, ctx)?);
        }

        // SAFETY: we initialized each element above by preading them out, correctness
        // of the initialized element is guaranted by pread itself
        Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
    }
}
```

i've also written a custom type with a custom drop impl and verified drop is only run once for each element (after the array leaves scope) whereas with the zeroed version, it is run twice (the `*element` deref as i mentioned)

Comment (Frostie314159) [reply src/ctx.rs:791]: The array init looks pretty good to me, my biggest worry is `Drop` not being called, since `T` could potentially and very realistically contain heap allocated objects like `Box` or `Vec` which wouldn't be freed properly.

Comment (m4b) [reply src/ctx.rs:791]: And here is a version that drops the elements if we fail while preading; also for some reason i can't get transmute working, which should be a free cast, according to MaybeUninit docs, i think perhaps it doesn't work with const generics (yet)?

```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        use std::mem::MaybeUninit;

        let mut buf: [MaybeUninit<T>; N] = std::array::from_fn(|_| MaybeUninit::uninit());

        let mut error_ctx = None;
        for (idx, element) in buf.iter_mut().enumerate() {
            match src.gread_with::<T>(&mut offset, ctx) {
                Ok(val) => {
                    *element = MaybeUninit::new(val);
                }
                Err(e) => {
                    error_ctx = Some((e, idx));
                }
            }
        }
        if let Some((e, idx)) = error_ctx {
            for element in &mut buf[0..idx].iter_mut() {
                unsafe {
                    element.assume_init_drop();
                }
            }
            Err(e)
        } else {
            // SAFETY: we initialized each element above by preading them out, correctness
            // of the initialized element is guaranted by pread itself
            Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
        }
    }
}
```

Comment (Frostie314159) [reply src/ctx.rs:791]: I had trouble with `transmute` to, which is why I used `[T; N]::map` instead, but this map will with certainty be optimised out, since the function calls doesn't actually do anything. So far this looks great.

**Comment by m4b** [comment]
I've never seen this error before, looks like transmute doesn't work with const generics, which is a bit sad:
```
error[E0512]: cannot transmute between types of different sizes, or dependently-sized types
   --> src/ctx.rs:815:26
    |
815 |             Ok((unsafe { std::mem::transmute::<_, Self>(buf) }, offset))
    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: source type: `[MaybeUninit<T>; N]` (this type does not have a fixed size)
    = note: target type: `[T; N]` (this type does not have a fixed size)
```

File: src/ctx.rs
Comment (Frostie314159) [reply src/ctx.rs:791]: The only thing that is missing, is that the first loop doesn't break on an error.

Comment (Frostie314159) [reply src/ctx.rs:791]: I just added that and have this impl now:
```rust
impl<'a, Ctx: Copy, T: TryFromCtx<'a, Ctx, Error = error::Error>, const N: usize>
    TryFromCtx<'a, Ctx> for [T; N]
{
    type Error = error::Error;
    fn try_from_ctx(src: &'a [u8], ctx: Ctx) -> Result<(Self, usize), Self::Error> {
        let mut offset = 0;

        let mut buf: [MaybeUninit<T>; N] = core::array::from_fn(|_| MaybeUninit::uninit());

        let mut error_ctx = None;
        for (idx, element) in buf.iter_mut().enumerate() {
            match src.gread_with::<T>(&mut offset, ctx) {
                Ok(val) => {
                    *element = MaybeUninit::new(val);
                }
                Err(e) => {
                    error_ctx = Some((e, idx));
                    break;
                }
            }
        }
        if let Some((e, idx)) = error_ctx {
            for element in &mut buf[0..idx].iter_mut() {
                unsafe {
                    element.assume_init_drop();
                }
            }
            Err(e)
        } else {
            // SAFETY: we initialized each element above by preading them out, correctness
            // of the initialized element is guaranted by pread itself
            Ok((buf.map(|element| unsafe { element.assume_init() }), offset))
        }

    }
}
```

Comment (m4b) [reply src/ctx.rs:791]: yes push this; good catch on that `break` too, that would have been a serious error! :)

Comment (Frostie314159) [reply src/ctx.rs]: I just pushed it, kudos for spotting the unsoundness.

**Comment by m4b** [comment]
btw i've been testing with this code, helpful to see where drops occur:
```rust
#[derive(Debug)]
struct StrDrop<'a> {
    #[allow(unused)]
    s: &'a str,
}

impl<'a> scroll::ctx::TryFromCtx<'a, scroll::Endian> for StrDrop<'a> {
    type Error = scroll::Error;
    fn try_from_ctx(
        bytes: &'a [u8],
        _ctx: scroll::Endian,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        let s = bytes.gread(&mut offset)?;
        Ok((Self { s }, offset))
    }
}

impl<'a> Drop for StrDrop<'a> {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
    }
}

#[derive(Debug)]
struct StringDrop {
    #[allow(unused)]
    s: String,
}

impl<'a> scroll::ctx::TryFromCtx<'a, scroll::Endian> for StringDrop {
    type Error = scroll::Error;
    fn try_from_ctx(
        bytes: &'a [u8],
        _ctx: scroll::Endian,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        let s: &str = bytes.gread(&mut offset)?;
        Ok((Self { s: String::from(s) }, offset))
    }
}

impl Drop for StringDrop {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
        drop(&mut self.s);
    }
}

#[test]
fn test_fixed_array_str() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41]);
    let res = bytes.pread_with::<[StrDrop; 3]>(0, Default::default());
    println!("{res:?}");
    assert!(false);
}

#[test]
fn test_fixed_array_string() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41, 0x0]);
    let res = bytes.pread_with::<[StringDrop; 3]>(0, Default::default());
    println!("{res:?}");
    assert!(false);
}
```

what bothers me is why the `[StringDrop; 3]` is not an erro, but `[StrDrop; 2]` is; need to figure out what's going on there, probably related to the slice logic that changed slightly for pread

**Comment by Frostie314159** [comment]
I also found a [thread](https://users.rust-lang.org/t/transmuting-a-generic-array/45645), which discusses the problem with transmute.

**Comment by Frostie314159** [comment]
I think I found the error, the byte array for `text_fixed_array_str` is missing one zero byte.

File: tests/api.rs
Comment (m4b) [inline tests/api.rs:273]: technically i think you need to call `drop(&mut self.s)`

File: tests/api.rs
Comment (m4b) [inline tests/api.rs:283]: i need to understand what's going on here, i'd expect this to panic with an array of 2, since we don't have null terminating byte, and that's the str pread default ctx iirc

File: tests/api.rs
Comment (Frostie314159) [reply tests/api.rs:273]: I removed it, since the compiler emitted a warning stating, that dropping a reference doesn't actually do anything.

File: tests/api.rs
Comment (Frostie314159) [reply tests/api.rs:283]: Both of these tests do panic, but since that is the expected behavior, I added `#[should_panic]` to the test.

File: tests/api.rs
Comment (Frostie314159) [reply tests/api.rs:273]: Here is the relevant doc link: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#dropping-references

File: tests/api.rs
Comment (m4b) [reply tests/api.rs:283]: But their length is 3.  I’d expect them to panic at 2. Which they do not

Comment (Frostie314159) [reply tests/api.rs:283]: Ah I see what you mean, I'm currently on mobile, but will look into this in an hour or so.

Comment (Frostie314159) [reply tests/api.rs:283]: Ok, I looked into it and ran a test, with this:
```rust
#[test]
fn string_test() {
    use scroll::Pread;
    let bytes = [0x45, 0x42, 0x0, 0x45, 0x41];
    let mut offset = 0;
    bytes.gread_with::<&str>(&mut offset, StrCtx::Delimiter(0)).unwrap();
    bytes.gread_with::<&str>(&mut offset, StrCtx::Delimiter(0)).unwrap();
}
```
This is an issue with the `TryFromCtx` impl on `&str`, it doesn't return an error if no null bytes is encountered, but instead creates a string from as many bytes as are present. This is unrelated to this PR and needs to be fixed separately.

Comment (Frostie314159) [reply tests/api.rs:283]: I opened a PR fixing this #96.

Comment (m4b) [reply tests/api.rs:283]: I don't think we should block this PR on that issue; adding generics is useful enough, and the only thing that's weird is the test; you can just remove the str based tests, though it would be useful to have an allocated, non fixed size/context dependent type (String was for this); I suggest this:

```rust
type MyCtx = ();

#[derive(Debug)]
struct BytesDrop<'a> {
    #[allow(unused)]
    s: &'a [u8],
}

impl<'a> scroll::ctx::TryFromCtx<'a, MyCtx> for BytesDrop<'a> {
    type Error = scroll::Error;
    fn try_from_ctx(
        src: &'a [u8],
        _ctx: MyCtx,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let len = src.iter().take_while(|c| **c != 0x0).count();
        if len > src.len() {
            return Err(scroll::Error::TooBig {
                size: len,
                len: src.len(),
            });
        };
        Ok((Self { s: &src[..len] }, len))
    }
}

impl<'a> Drop for BytesDrop<'a> {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
    }
}

#[derive(Debug)]
struct VecDrop {
    #[allow(unused)]
    s: Vec<u8>,
}

impl<'a> scroll::ctx::TryFromCtx<'a, MyCtx> for VecDrop {
    type Error = scroll::Error;
    fn try_from_ctx(
        bytes: &'a [u8],
        _ctx: MyCtx,
    ) -> ::std::result::Result<(Self, usize), Self::Error> {
        let mut offset = 0;
        let s: BytesDrop = bytes.gread(&mut offset)?;
        Ok((Self { s: Vec::from(s.s) }, offset))
    }
}

impl Drop for VecDrop {
    fn drop(&mut self) {
        println!("Dropping {self:?}");
        drop(&mut self.s);
    }
}

#[test]
fn test_fixed_array_str() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41]);
    let res = bytes.pread_with::<[BytesDrop; 2]>(0, Default::default());
    println!("{res:?}");
    assert!(res.is_ok());
}

#[test]
fn test_fixed_array_string() {
    use scroll::Pread;
    let bytes = Box::new([0x45, 0x42, 0x0, 0x45, 0x41, 0x0]);
    let res = bytes.pread_with::<[VecDrop; 2]>(0, Default::default());
    println!("{res:?}");
    assert!(res.is_ok());
}
```

can add check that the inner bytes of the BytesDrop and VecDrop are as expected, if want. After that, let's merge this!

Comment (Frostie314159) [reply tests/api.rs]: I just pushed this. Blocking this will probably not be necessary.

File: tests/api.rs
Comment (m4b) [inline tests/api.rs:287]: you didn't need to actually box these bytes but it's fine, doesn't matter

Comment (Frostie314159) [reply tests/api.rs:287]: I actually just copied the tests, but I'm gonna resolve this quickly since it's unnecessary.

**Review by m4b — APPROVED** [review]

File: tests/api.rs
Comment (Frostie314159) [reply tests/api.rs]: Fixed with latest commit.

File: tests/api.rs
Comment (m4b) [reply tests/api.rs:280]: afaik this is a leak in our VecDrop, since you aren't dropping the owned Vec here. The above would apply to the BytesDrop, which is a reference, as you say, which doesn't need to be dropped technically. Since this is a test it doesn't really matter and can be fixed in a followup, but i'm very sure this is a memory leak in this drop impl.

Comment (Frostie314159) [reply tests/api.rs:280]: Actually, the fields of the struct are still dropped by the compiler, see: https://doc.rust-lang.org/std/ops/trait.Drop.html#examples

Comment (m4b) [reply tests/api.rs:280]: ah thank you, I will give it a read, haven't written drop explicitly in a long while, so good to know :)

**Comment by m4b** [comment]
thank you for your patience @Frostie314159 , and thanks for the PR!

**Comment by Frostie314159** [comment]
No problem, the patience was definitely worth it, since we now have a sound implementation.

**Comment by Frostie314159** [comment]
@m4b I just noticed, in one of my projects, that LLVM fails to optimize this properly, when writing byte arrays. We should probably add a warning about that, since it crippled the performance in my case.

**Comment by m4b** [comment]
@Frostie314159 thanks for noting this; there is another issue about llvm failing to optimize before, I wonder if it's related? let me find it; is it worth investigating filing this upstream to rust about perf issue ?

**Comment by m4b** [comment]
Here was the issue, I don't know if it's relevant anymore, it was years ago: https://github.com/m4b/scroll/issues/53

can you paste the assembly if possible, or why you believe the fixed size array is failing to optimize well?

**Pull request merged**
